### PR TITLE
Serialize std::optional and std::variant

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

Add serialization for std::optional and std::variant

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/Parallel/PupStlCpp17.hpp [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    if (valid) {
+      p | *t;
+    }
+  }
+}
+
+/// \ingroup ParallelGroup
+/// Serialization of std::optional for Charm++
+template <typename T>
+void operator|(er& p, std::optional<T>& t) noexcept {  // NOLINT
```
Comment (geoffrey4444) [inline src/Parallel/PupStlCpp17.hpp:41]: Why is this just er& instead of PUP::er&?

File: src/Parallel/PupStlCpp17.hpp [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  assert(not t.valueless_by_exception());  // NOLINT
+
+  size_t current_index = 0;
+  size_t send_index = t.index();
+  p | send_index;
+
+  const auto pup_helper = [&current_index, &p, send_index,
+                           &t](auto type_v) noexcept {
+    using type = tmpl::type_from<decltype(type_v)>;
+    if (current_index == send_index) {
```
Comment (geoffrey4444) [inline src/Parallel/PupStlCpp17.hpp]: This is false every time but one... use if UNLIKELY?

File: tests/Unit/Parallel/Test_PupStlCpp17.cpp [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  {
+    INFO("Optional");
+    const std::optional<int> t_not_set{};
+    CHECK_FALSE(t_not_set.has_value());
+    const auto t_not_set_deserialized = serialize_and_deserialize(t_not_set);
+    CHECK_FALSE(t_not_set_deserialized.has_value());
+
+    const std::optional<int> t_set{10};
+    REQUIRE(t_set.value());
+    CHECK(*t_set == 10);
```
Comment (geoffrey4444) [inline tests/Unit/Parallel/Test_PupStlCpp17.cpp]: How could this ever fail?

**Review by geoffrey4444 — CHANGES_REQUESTED** [review]
Just a couple of quick questions / suggestions

File: src/Parallel/PupStlCpp17.hpp (hunk 1 above)
Comment (nilsdeppe) [reply src/Parallel/PupStlCpp17.hpp:41]: Because this function is added to the `PUP` namespace

Comment (geoffrey4444) [reply src/Parallel/PupStlCpp17.hpp:41]: ... but then why do you say pup::er& on line 76 in the analogous call frosted::variant? That looks to me like it's also in the PUP namespace

Comment (nilsdeppe) [reply src/Parallel/PupStlCpp17.hpp:41]: Probably some historical thing about how I wrote the code. It should be removed. I will remove it :) Thanks!

File: src/Parallel/PupStlCpp17.hpp [hunk 4]
```diff
@@ -37,4 +40,40 @@ template <typename T>
...
+      if (p.isUnpacking()) {
+        type temp{};
+        p | temp;
+        t = std::move(temp);
+      } else {
+        p | std::get<type>(t);
+      }
+    }
+    current_index++;
```
Comment (moxcodes) [inline src/Parallel/PupStlCpp17.hpp]: [very nitpick] prefix operator maybe.

**Review by moxcodes — DISMISSED** [review]
Looks good, go ahead and squash as soon as @geoffrey4444  gives the go ahead (along with the ++ if you care to change it).
Cheers!

**Comment by geoffrey4444** [comment]
Please squash

**Comment by nilsdeppe** [comment]
Okay, squashed. Thanks for the reviews! :)

**Review by moxcodes — APPROVED** [review]

**Review by geoffrey4444 — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Pull request merged**